src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/ComponentAccessibilityCodeActionProvider.cs

```diff
@@ -0,0 +1,219 @@
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    internal class ComponentAccessibilityCodeActionProvider : RazorCodeActionProvider
+    {
+        private readonly TagHelperFactsService _tagHelperFactsService;
+        private readonly FilePathNormalizer _filePathNormalizer;
+
+        public ComponentAccessibilityCodeActionProvider(
+            TagHelperFactsService tagHelperFactsService,
+            FilePathNormalizer filePathNormalizer)
+        {
+            _tagHelperFactsService = tagHelperFactsService ?? throw new ArgumentNullException(nameof(tagHelperFactsService));
+            _filePathNormalizer = filePathNormalizer ?? throw new ArgumentNullException(nameof(filePathNormalizer));
+        }
+
+        override public Task<CommandOrCodeActionContainer> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            var commandOrCodeActions = new List<CommandOrCodeAction>();
+
+            // Locate cursor
+            var change = new SourceChange(context.Location.AbsoluteIndex, length: 0, newText: string.Empty);
+            var node = context.CodeDocument.GetSyntaxTree().Root.LocateOwner(change);
+            if (node is null)
+            {
+                return Task.FromResult<CommandOrCodeActionContainer>(null);
+            }
+
+            // Find start tag
+            var startTag = (MarkupStartTagSyntax)node.Ancestors().FirstOrDefault(n => n is MarkupStartTagSyntax);
+            if (startTag == null)
+            {
+                return Task.FromResult<CommandOrCodeActionContainer>(null);
+            }
+
+            // Ignore if has dots, as we only handle short tags
+            if (startTag.Name.Content.Contains("."))
+            {
+                return Task.FromResult<CommandOrCodeActionContainer>(null);
+            }
+
+            if (IsTagUnknown(startTag, context))
+            {
+                AddCreateComponentFromTag(context, startTag, commandOrCodeActions);
+                AddComponentAccessFromTag(context, startTag, commandOrCodeActions);
+            }
+
+            return Task.FromResult(new CommandOrCodeActionContainer(commandOrCodeActions));
+        }
+
+        private void AddCreateComponentFromTag(RazorCodeActionContext context, MarkupStartTagSyntax startTag, List<CommandOrCodeAction> container)
+        {
+            var path = _filePathNormalizer.Normalize(context.Request.TextDocument.Uri.GetAbsoluteOrUNCPath());
+            var newComponentPath = Path.Combine(Path.GetDirectoryName(path), $"{startTag.Name.Content}.razor");
+            if (File.Exists(newComponentPath))
+            {
+                return;
+            }
+
+            var actionParams = new CreateComponentCodeActionParams
+            {
+                Uri = context.Request.TextDocument.Uri,
+                Path = newComponentPath,
+            };
+            var data = JObject.FromObject(actionParams);
```

Comment:
The problem here is that we want data to be polymorphic, in that we always use the first index in the array and just pass an object with named arguments that the corresponding resolver will understand. As it stands I don't know a better way to do this, but I could write up what's going on if we want to take a more comprehensive look.

Reply:
Can you please provide a link to where the resolver consumes this?

Reply:
https://github.com/dotnet/aspnetcore-tooling/blob/de2e4c6b90d0b5e2e45e0cdad461985c8d4a0c49/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionResolver.cs#L43

Reply:
https://github.com/dotnet/aspnetcore-tooling/blob/de2e4c6b90d0b5e2e45e0cdad461985c8d4a0c49/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CreateComponentCodeActionResolver.cs#L42

Reply:
This should be possible using a base class and some basic `JsonProperty`/converter. 

It would be de-serialized like:
https://github.com/dotnet/aspnetcore-tooling/blob/015feb105fe92b0b8d4191b02da84e779476ca9c/src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Serialization/TagHelperResolutionResultSerializationTest.cs#L65

Let's discuss this offline when you have a moment, I don't want to block this PR.